File: backend/server/src/main/kotlin/com/jervis/infrastructure/llm/DocumentExtractionClient.kt
package com.jervis.infrastructure.llm

import com.google.protobuf.ByteString
import com.jervis.contracts.common.RequestContext
import com.jervis.contracts.common.Scope
import com.jervis.contracts.document_extraction.DocumentExtractionServiceGrpcKt
import com.jervis.contracts.document_extraction.ExtractRequest
import io.grpc.ManagedChannel
import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder
import mu.KotlinLogging
import java.util.UUID
import java.util.concurrent.TimeUnit

private val logger = KotlinLogging.logger {}

/**
 * gRPC client for the Python document-extraction microservice.
 *
 * Replaces the former REST client on `POST /extract-base64`. The base
 * URL is still expressed as "host[:port]" (legacy config shape) — port
 * 5501 is the fixed gRPC target on the document-extraction pod.
 */
class DocumentExtractionClient(baseUrl: String) {

    private val channel: ManagedChannel = run {
        val cleaned = baseUrl.trim().trimEnd('/')
        val hostPart = if ("://" in cleaned) cleaned.substringAfter("://") else cleaned
        val host = hostPart.substringBefore('/').substringBefore(':')
        NettyChannelBuilder.forAddress(host, 5501)
            .usePlaintext()
            .maxInboundMessageSize(64 * 1024 * 1024)
            .keepAliveTime(30, TimeUnit.SECONDS)
            .keepAliveTimeout(5, TimeUnit.SECONDS)
            .keepAliveWithoutCalls(true)
            .build()
    }

    private val stub = DocumentExtractionServiceGrpcKt
        .DocumentExtractionServiceCoroutineStub(channel)

    private fun ctx(): RequestContext =
        RequestContext.newBuilder()
            .setScope(Scope.newBuilder().build())
            .setRequestId(UUID.randomUUID().toString())
            .setIssuedAtUnixMs(System.currentTimeMillis())
            .build()

    suspend fun extractText(content: String, mimeType: String = "text/html"): String {
        if (content.isBlank()) return content

        val filename = when {
            mimeType.contains("html") -> "content.html"
            mimeType.contains("xml") -> "content.xml"
            else -> "content.txt"
        }

        val resp = stub.extract(
            ExtractRequest.newBuilder()
                .setCtx(ctx())
                .setContent(ByteString.copyFrom(content.toByteArray(Charsets.UTF_8)))
                .setFilename(filename)
                .setMimeType(mimeType)
                .setMaxTier("NONE")
                .build(),
        )
        logger.info { "Document extraction: ${content.length} chars → ${resp.text.length} chars (method=${resp.method})" }
        return resp.text
    }

    suspend fun extractBytes(
        fileBytes: ByteArray,
        filename: String,
        mimeType: String,
        maxTier: String = "NONE",
    ): String {
        val resp = stub.extract(
            ExtractRequest.newBuilder()
                .setCtx(ctx())
                .setContent(ByteString.copyFrom(fileBytes))
                .setFilename(filename)
                .setMimeType(mimeType)
                .setMaxTier(maxTier)
                .build(),
        )
        logger.info { "Document extraction: $filename (${fileBytes.size} bytes) → ${resp.text.length} chars (method=${resp.method})" }
        return resp.text
    }

    fun shutdown() {
        runCatching { channel.shutdown().awaitTermination(5, TimeUnit.SECONDS) }
    }
}

File: backend/service-orchestrator/app/document_extraction_client.py
"""gRPC client for the Python document-extraction microservice."""

from __future__ import annotations

import logging
from typing import Optional

import grpc.aio

from app.config import settings
from jervis.common import types_pb2
from jervis.document_extraction import extract_pb2, extract_pb2_grpc

logger = logging.getLogger("orchestrator.document_extraction")

_channel: Optional[grpc.aio.Channel] = None
_stub: Optional[extract_pb2_grpc.DocumentExtractionServiceStub] = None


def _target() -> str:
    url = (
        getattr(settings, "document_extraction_url", None)
        or "http://jervis-document-extraction:8080"
    )
    url = url.rstrip("/")
    if "://" in url:
        url = url.split("://", 1)[1]
    host = url.split("/")[0].split(":")[0]
    return f"{host}:5501"


def _get_stub() -> extract_pb2_grpc.DocumentExtractionServiceStub:
    global _channel, _stub
    if _stub is None:
        _channel = grpc.aio.insecure_channel(
            _target(),
            options=[
                ("grpc.max_receive_message_length", 64 * 1024 * 1024),
                ("grpc.max_send_message_length", 64 * 1024 * 1024),
            ],
        )
        _stub = extract_pb2_grpc.DocumentExtractionServiceStub(_channel)
    return _stub


def _ctx() -> types_pb2.RequestContext:
    return types_pb2.RequestContext(
        request_id="",
        trace={"caller": "service-orchestrator"},
    )


async def document_extraction_extract(
    content: bytes,
    filename: str,
    mime_type: str = "",
    max_tier: str = "NONE",
    timeout: float = 120.0,
) -> dict:
    stub = _get_stub()
    resp = await stub.Extract(
        extract_pb2.ExtractRequest(
            ctx=_ctx(),
            content=bytes(content),
            filename=filename,
            mime_type=mime_type,
            max_tier=max_tier,
        ),
        timeout=timeout,
    )
    return {
        "text": resp.text,
        "method": resp.method,
        "metadata": dict(resp.metadata),
        "pages": [
            {"page_number": p.page_number, "text": p.text}
            for p in resp.pages
        ],
    }
